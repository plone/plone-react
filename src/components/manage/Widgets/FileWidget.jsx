/**
 * FileWidget component.
 * @module components/manage/Widgets/FileWidget
 */

import React from 'react';
import PropTypes from 'prop-types';
import { Form, Grid, Input, Label, Button } from 'semantic-ui-react';
import { map } from 'lodash';
import { readAsDataURL } from 'promise-file-reader';

import deleteSVG from '../../../icons/delete.svg';
import { Icon } from '../../../components';

/**
 * FileWidget component class.
 * @function FileWidget
 * @returns {string} Markup of the component.
 */
const FileWidget = ({
  id,
  title,
  required,
  description,
  error,
  value,
  onChange,
  fieldSet,
}) => {
  const fileInput = React.useRef(null);

  return (
    <Form.Field
      inline
      required={required}
      error={error.length > 0}
      className={description ? 'help' : ''}
      id={`${fieldSet || 'field'}-${id}`}
    >
      <Grid>
        <Grid.Row stretched>
          <Grid.Column width="4">
            <div className="wrapper">
              <label htmlFor={`field-${id}`}>{title}</label>
            </div>
          </Grid.Column>
          <Grid.Column width="8">
            <Input
              id={`field-${id}`}
              name={id}
              type="file"
              ref={fileInput}
              onChange={({ target }) => {
                const file = target.files[0];
                readAsDataURL(file).then(data => {
                  const fields = data.match(/^data:(.*);(.*),(.*)$/);
                  onChange(id, {
                    data: fields[3],
                    encoding: fields[2],
                    'content-type': fields[1],
                    filename: file.name,
                  });
                });
              }}
            />
            <div className="field-file-name">
              {value && value.filename}
              {value && (
                <Button
                  icon
                  basic
                  className="delete-button"
                  aria-label="delete file"
                  onClick={() => {
                    onChange(id, null);
                    fileInput.current.inputRef.value = null;
                  }}
                >
                  <Icon name={deleteSVG} size="20px" />
                </Button>
              )}
            </div>
            {map(error, message => (
              <Label key={message} basic color="red" pointing>
                {message}
              </Label>
            ))}
          </Grid.Column>
        </Grid.Row>
        {description && (
          <Grid.Row stretched>
            <Grid.Column stretched width="12">
              <p className="help">{description}</p>
            </Grid.Column>
          </Grid.Row>
        )}
      </Grid>
    </Form.Field>
  );
};

/**
 * Property types.
 * @property {Object} propTypes Property types.
 * @static
 */
FileWidget.propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  description: PropTypes.string,
  required: PropTypes.bool,
  error: PropTypes.arrayOf(PropTypes.string),
  value: PropTypes.shape({
    '@type': PropTypes.string,
    title: PropTypes.string,
  }),
  onChange: PropTypes.func.isRequired,
};

/**
 * Default properties.
 * @property {Object} defaultProps Default properties.
 * @static
 */
FileWidget.defaultProps = {
  description: null,
  required: false,
  error: [],
  value: null,
};

export default FileWidget;
