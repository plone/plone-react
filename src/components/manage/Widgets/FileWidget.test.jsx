import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import FileWidget from './FileWidget';
import { Provider } from 'react-intl-redux';

const mockStore = configureStore();

test('renders a file widget component', () => {
  const store = mockStore({
    intl: {
      locale: 'en',
      messages: {},
    },
  });

  const component = renderer.create(
    <Provider store={store}>
      <FileWidget id="my-field" title="My field" onChange={() => {}} />,
    </Provider>,
  );
  const json = component.toJSON();
  expect(json).toMatchSnapshot();
});
