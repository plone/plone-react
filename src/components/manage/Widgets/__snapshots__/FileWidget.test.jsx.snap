// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a file widget component 1`] = `
Array [
  <div
    className="inline field"
    id="field-my-field"
  >
    <div
      className="ui grid"
    >
      <div
        className="stretched row"
      >
        <div
          className="four wide column"
        >
          <div
            className="wrapper"
          >
            <label
              htmlFor="field-my-field"
            >
              My field
            </label>
          </div>
        </div>
        <div
          className="eight wide column"
        >
          <div
            className="file-widget-dropzone"
            onBlur={[Function]}
            onClick={[Function]}
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            tabIndex={0}
          >
            <div
              className="dropzone-placeholder"
            >
              <p
                className="dropzone-text"
              >
                Drag the new item to upload
              </p>
            </div>
            <label
              className="label-file-widget-input"
            >
              Choose a File/Image
            </label>
            <input
              autoComplete="off"
              id="field-my-field"
              multiple={true}
              name="my-field"
              onChange={[Function]}
              onClick={[Function]}
              style={
                Object {
                  "display": "none",
                }
              }
              tabIndex={-1}
              type="file"
            />
          </div>
          <div
            className="field-file-name"
          />
        </div>
      </div>
    </div>
  </div>,
  ",",
]
`;
